src/analysis/normed_space/compact_operator.lean

```diff
@@ -0,0 +1,375 @@
+  [topological_space M₃] [add_comm_monoid M₁] [module R₁ M₁]
+
+lemma is_compact_operator.comp_clm [add_comm_monoid M₂] [module R₂ M₂] {f : M₂ → M₃}
+  (hf : is_compact_operator f) (g : M₁ →SL[σ₁₂] M₂) :
+  is_compact_operator (f ∘ g) :=
+begin
+  have := g.continuous.tendsto 0,
+  rw map_zero at this,
+  rcases hf with ⟨K, hK, hKf⟩,
+  exact ⟨K, hK, this hKf⟩
+end
+
+lemma is_compact_operator.continuous_comp {f : M₁ → M₂} (hf : is_compact_operator f) {g : M₂ → M₃}
+  (hg : continuous g) :
+  is_compact_operator (g ∘ f) :=
+begin
+  rcases hf with ⟨K, hK, hKf⟩,
+  refine ⟨g '' K, hK.image hg, mem_of_superset hKf _⟩,
+  nth_rewrite 1 preimage_comp,
+  exact preimage_mono (subset_preimage_image _ _)
+end
+
+lemma is_compact_operator.clm_comp [add_comm_monoid M₂] [module R₂ M₂] [add_comm_monoid M₃]
+  [module R₃ M₃] {f : M₁ → M₂} (hf : is_compact_operator f) (g : M₂ →SL[σ₂₃] M₃) :
+  is_compact_operator (g ∘ f) :=
+hf.continuous_comp g.continuous
+
+lemma is_compact_operator.continuous_on_comp [t2_space M₂] {f : M₁ → M₂}
+  (hf : is_compact_operator f) {g : M₂ → M₃} (hg : continuous_on g (closure $ range f)) :
+  is_compact_operator (g ∘ f) :=
+begin
+  rw is_compact_operator_iff_exists_mem_nhds_is_compact_closure_image at hf,
+  rcases hf with ⟨V, hV, hVf⟩,
+  refine ⟨g '' (closure $ f '' V),
+          hVf.image_of_continuous_on (hg.mono $ closure_mono $ image_subset_range _ _),
+          mem_of_superset hV _⟩,
+  calc
+    V ⊆ f ⁻¹' (f '' V) : subset_preimage_image _ _
+  ... ⊆ f ⁻¹' (closure $ f '' V) : preimage_mono subset_closure
+  ... ⊆ f ⁻¹' (g ⁻¹' (g '' (closure $ f '' V))) : preimage_mono (subset_preimage_image _ _)
+end
+
+end comp
+
+section cod_restrict
+
+variables {R₁ R₂ : Type*} [semiring R₁] [semiring R₂] {σ₁₂ : R₁ →+* R₂}
+  {M₁ M₂ : Type*} [topological_space M₁] [topological_space M₂]
+  [add_comm_monoid M₁] [add_comm_monoid M₂] [module R₁ M₁] [module R₂ M₂]
+
+lemma is_compact_operator.cod_restrict {f : M₁ → M₂} (hf : is_compact_operator f)
+  {V : submodule R₂ M₂} (hV : ∀ x, f x ∈ V) (h_closed : is_closed (V : set M₂)):
+  is_compact_operator (set.cod_restrict f V hV) :=
+let ⟨K, hK, hKf⟩ := hf in
+⟨coe ⁻¹' K, (closed_embedding_subtype_coe h_closed).is_compact_preimage hK, hKf⟩
+
+end cod_restrict
+
+section restrict_invariant
+
+variables {R₁ R₂ R₃ : Type*} [semiring R₁] [semiring R₂] [semiring R₃] {σ₁₂ : R₁ →+* R₂}
+  {σ₂₃ : R₂ →+* R₃} {M₁ M₂ M₃ : Type*} [topological_space M₁] [uniform_space M₂]
+  [topological_space M₃] [add_comm_monoid M₁] [add_comm_monoid M₂] [add_comm_monoid M₃]
+  [module R₁ M₁] [module R₂ M₂] [module R₃ M₃]
+
+/-- If a compact operator preserves a closed submodule, its restriction to that submodule is
+compact. -/
+lemma is_compact_operator.restrict_invariant {f : M₁ →ₗ[R₁] M₁} (hf : is_compact_operator f)
+  {V : submodule R₁ M₁} (hV : ∀ v ∈ V, f v ∈ V) (h_closed : is_closed (V : set M₁)):
+  is_compact_operator (f.restrict hV) :=
```

Comment:
By analogy with `linear_map`, this should apparently just be called `restrict`, not `restrict_invariant`.  The last time this came up (also in the context of eigenvalue theory), there was a discussion with @abentkamp and @Vierkantor https://github.com/leanprover-community/mathlib/pull/4053#discussion_r484342212 about whether one should allow for more generality -- different domain and range.  The argument against generality is from @abentkamp 

> For endomorphisms, restricting both at once makes more sense because you put an endomorphism in and you get an endomorphism out.

which I think for now is still reasonable.  But it'd be good to make the docstring on this lemma crystal clear that `restrict` is only an operation on endomorphisms.  (Maybe even explain the recipe in the docstring for dealing with more general settings: combine `cod_restrict` and composition with `subtypeL`.)

Reply:
I like the expanded docstring.  Another point about this lemma (and the next): can you semilinearize it?

Reply:
Well if I want to stick with `linear_map.restrict` I guess not